packages/expect/src/types.ts

```diff
@@ -20,16 +20,29 @@ export type AsyncExpectationResult = Promise<SyncExpectationResult>;
 
 export type ExpectationResult = SyncExpectationResult | AsyncExpectationResult;
 
+export type MatcherFunctionWithContext<
+  Context extends MatcherState = MatcherState,
+  Expected extends Array<any> = [],
+> = (
+  this: Context,
+  actual: unknown,
+  ...expected: Expected
+) => ExpectationResult;
+
+export type MatcherFunction<Expected extends Array<any> = []> =
+  MatcherFunctionWithContext<MatcherState, Expected>;
+
+// TODO should be replaced with `MatcherFunctionWithContext`
```

Comment:
Exposed types look good and have tests, but I have to figure out if and how to use it internally.